datafusion-examples/examples/flight_sql_server.rs

```diff
@@ -0,0 +1,336 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow_flight::sql::SqlInfo;
+use arrow_flight::{FlightEndpoint, SchemaAsIpc};
+use datafusion::datasource::file_format::parquet::ParquetFormat;
+use datafusion::datasource::listing::ListingOptions;
+use datafusion::datasource::object_store::local::LocalFileSystem;
+use tonic::transport::Server;
+use tonic::{Response, Status};
+
+use datafusion::prelude::*;
+
+use arrow_flight::{
+    flight_service_server::FlightService,
+    flight_service_server::FlightServiceServer,
+    sql::{
+        server::FlightSqlService, ActionClosePreparedStatementRequest,
+        ActionCreatePreparedStatementRequest, CommandGetCatalogs,
+        CommandGetCrossReference, CommandGetDbSchemas, CommandGetExportedKeys,
+        CommandGetImportedKeys, CommandGetPrimaryKeys, CommandGetSqlInfo,
+        CommandGetTableTypes, CommandGetTables, CommandPreparedStatementQuery,
+        CommandPreparedStatementUpdate, CommandStatementQuery, CommandStatementUpdate,
+        ProstMessageExt, TicketStatementQuery,
+    },
+    FlightData, FlightDescriptor, FlightInfo, Ticket,
+};
+use prost::Message;
+
+#[derive(Clone)]
+pub struct FlightSqlServiceImpl {
+    parquet_file: String,
+}
+
+impl FlightSqlServiceImpl {
+    fn new() -> Self {
+        let testdata = datafusion::arrow::util::test_util::parquet_test_data();
+        FlightSqlServiceImpl {
+            parquet_file: format!("{}/alltypes_plain.parquet", testdata),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl FlightSqlService for FlightSqlServiceImpl {
+    type FlightService = FlightSqlServiceImpl;
+    // get_flight_info
+    async fn get_flight_info_statement(
+        &self,
+        query: CommandStatementQuery,
+        request: FlightDescriptor,
+    ) -> Result<Response<FlightInfo>, Status> {
+        let listing_options = ListingOptions::new(Arc::new(ParquetFormat::default()));
+
+        let schema = listing_options
+            .infer_schema(Arc::new(LocalFileSystem {}), &self.parquet_file)
+            .await
+            .unwrap();
+
+        let ticket = TicketStatementQuery {
+            statement_handle: query.query.into_bytes(),
+        };
+        let endpoint = FlightEndpoint {
+            ticket: Some(Ticket {
+                ticket: ticket.as_any().encode_to_vec(),
+            }),
+            location: vec![],
+        };
+        let info = FlightInfo {
```

Comment:
It seems like the `FlightInfo` should be returned as part of the call that is executing the query in `do_get_statement` -- maybe even handled by the implementation of `FlightSqlService`

Though I realize it may be somewhat tricky to save the state between what is essentially two calls in flight -- the DoPut (which returns this FlightInfo) and then the `GetFlightInfo`)

Reply:
I don't quite get it. Are you suggesting we merge the get_flight_info_statement and do_get_statement?

No we can't do that, we need to leave that to the client, that's part of the protocol. 

You may have noticed the endpoint field, when we implement a distributed sql server, we can have a coordinator to serve the get_flight_info_statement, and then the client can be redirected to connected to multiple endpoints returned by the coordinator, so that not all data need go through one server host. That's just one case we can think of.